src/crypto/noise/mod.rs

```diff
@@ -656,17 +670,35 @@ impl<S: AsyncRead + AsyncWrite + Unpin> AsyncWrite for NoiseSocket<S> {
     }
 }
 
-/// Try to parse `PeerId` from received `NoiseHandshakePayload`
-fn parse_peer_id(buf: &[u8]) -> Result<PeerId, NegotiationError> {
-    match handshake_schema::NoiseHandshakePayload::decode(buf) {
-        Ok(payload) => {
-            let identity = payload.identity_key.ok_or(NegotiationError::PeerIdMissing)?;
+/// Parse the `PeerId` from received `NoiseHandshakePayload` and verify the payload signature.
+fn parse_and_verify_peer_id(
+    payload: handshake_schema::NoiseHandshakePayload,
+    keypair_public: &[u8],
+) -> Result<PeerId, NegotiationError> {
+    let identity = payload.identity_key.ok_or(NegotiationError::PeerIdMissing)?;
+    let remote_public_key = PublicKey::from_protobuf_encoding(&identity)?;
+    let remote_key_signature =
+        payload.identity_sig.ok_or(NegotiationError::BadSignature).map_err(|err| {
+            tracing::debug!(target: LOG_TARGET, "payload without signature");
+            err
+        })?;
+
+    let peer_id = PeerId::from_public_key(&remote_public_key);
+
+    if !remote_public_key.verify(
+        &[STATIC_KEY_DOMAIN.as_bytes(), keypair_public].concat(),
+        &remote_key_signature,
```

Comment:
I am a little bit confused: if `remote_public_key` is a network public key of a remote peer, then what is `keypair_public` and why we don't check them to be equal?

Reply:
Will have a look, we can include this in the next release 👍